src/FromData.purs

```diff
@@ -5,27 +5,109 @@ module FromData
+import Control.Alternative (guard)
+import Data.Array (cons, head, uncons, (:))
 import Data.Array as Array
 import Data.BigInt (BigInt)
+import Data.BigInt as BigInt
 import Data.Either (Either(Left, Right))
+import Data.Generic.Rep (repOf)
+import Data.Generic.Rep as G
 import Data.List (List)
 import Data.Map (Map)
 import Data.Map as Map
 import Data.Maybe (Maybe(Nothing, Just), maybe)
 import Data.Ratio (Ratio, reduce)
+import Data.Symbol (class IsSymbol, SProxy(..), reflectSymbol, reifySymbol)
 import Data.Traversable (for, traverse)
 import Data.Tuple (Tuple(Tuple))
 import Data.Tuple.Nested (type (/\), (/\))
 import Data.UInt (UInt)
 import Data.Unfoldable (class Unfoldable)
 import Helpers (bigIntToUInt)
 import Prim.TypeError (class Fail, Text)
+import Type.Proxy (Proxy(..))
 import Types.ByteArray (ByteArray)
 import Types.PlutusData (PlutusData(Bytes, Constr, List, Map, Integer))
 
+-- Classes
+
 class FromData (a :: Type) where
   fromData :: PlutusData -> Maybe a
 
+class ConstrIndex (a :: Type) where
+  constrIndex :: Proxy a -> Map Int String
+
+class FromDataWithIndex a ci where
+  fromDataWithIndex :: ConstrIndex ci => Proxy a -> Proxy ci -> PlutusData -> Maybe a
+
+class FromDataArgs a where
+  fromDataArgs :: Array PlutusData -> Maybe a
+
+-- Default constructor indices
+class CountedConstrIndex (a :: Type) where
+  countedConstrIndex :: Proxy a -> Int -> Map Int String -> Map Int String
+
+resolveConstr :: forall a. ConstrIndex a => Proxy a -> Int -> Maybe String
+resolveConstr pa i = Map.lookup i (constrIndex pa)
+
+-- instance FromDataArgs G.NoArguments where
+--   fromDataArgs pdArgs = case uncons pdArgs of
+--     Just {head: (Constr i []), tail: pds} -> Just G.NoArguments
+--     Nothing -> Nothing
+
+-- > data TestType = C0 | C1 Int | C2 Int String | C3 Int String Boolean
+-- > derive instance getTestType :: Generic TestType _
+-- > :t (from C0)
+-- Sum
+--  (Constructor "C0" NoArguments)
+--  (Sum
+--     (Constructor "C1" (Argument Int))
+--     (Sum
+--       (Constructor "C2"
+--         (Product (Argument Int) (Argument String)))
+--       (Constructor "C3"
+--         (Product
+--           (Argument Int)
+--           (Product
+--             (Argument String)
+--             (Argument Boolean))))))
+
+instance (ConstrIndex a, FromDataWithIndex l a, FromDataWithIndex r a) => FromDataWithIndex (G.Sum l r) a where
+  fromDataWithIndex _ pci pd = G.Inl <$> fromDataWithIndex (Proxy :: Proxy l) pci pd
+    <|> G.Inr <$> fromDataWithIndex (Proxy :: Proxy r) pci pd
+
+-- https://purescript-simple-json.readthedocs.io/en/latest/generics-rep.html
+instance (IsSymbol n, ConstrIndex a, FromDataArgs arg) => FromDataWithIndex (G.Constructor n arg) a where
+  fromDataWithIndex _ pci (Constr i pdArgs) = do
+    ix <- BigInt.toInt i
+    cn <- resolveConstr pci ix
+    let rn = reflectSymbol (Proxy :: Proxy n)
+    guard $ cn == rn
+    repArgs <- fromDataArgs pdArgs
+    pure $ (G.Constructor repArgs :: G.Constructor n arg)
```

Comment:
Is the explicit type sig for readability as it's inferred?

Reply:
There was likely a type problem and I tried things until the type checker turned green. It doesn't seem to be necessary so removing it